projects/common/src/telemetry/telemetry.ts

```diff
@@ -0,0 +1,43 @@
+import { InjectionToken } from '@angular/core';
+import { Dictionary } from './../utilities/types/types';
+
+export interface UserTelemetry {
+  readonly config: UserTelemetryConfig;
+  readonly initialized: boolean;
+  initialize(): void;
+  identify(userTraits: UserTraits): void;
+  trackEvent(name: string, eventData: Dictionary<unknown>): void;
+  trackPage(url: string): void;
+  shutdown(): void;
+}
+
+export interface UserTraits extends Dictionary<unknown> {
+  email: string;
+  companyName: string;
+  name: string;
+  displayName: string;
+}
+
+export interface UserTelemetryConfig<TConfig extends TelemetryProviderConfig = TelemetryProviderConfig> {
+  telemetryProvider: UserTelemetryType | string;
```

Comment:
how would the token approach work here for custom impl ? With injection token, sure I can use it for defining a custom module. But registration is happening at module load time only. So, I can just use that to register a given provider with the provider key and then do a look up when a config with matching key is passed. I won't be injecting it in a service. 

So, I made the key to accept a string. Some strings are ootb just like table cell renderers. But you also have an option to register and pass custom key.

Reply:
Table cell renderers are a slightly, but not completely, different story. We register them exactly as I'm describing - by providing a class. The string association is there because we want to support configuring the table in json where we only have strings - otherwise, I'd just define my columns with components - way easier! The unfortunate concusion there - we didn't really need to build that string->renderer mapping into table, we could have left it as the responsibility of the table widget only.

So as to how this would work - imagine I'm given something like
```ts
    UserTelemetryModule.withProviders([
      {
        telemetryProvider: FullStoryTelemetryProvider, // Instead of: UserTelemetryType.FullStory,
        telemetryProviderConfig: { orgId: environment.fullStoryOrgId },
        enablePageTracking: true,
        enableEventTracking: true,
        enableErrorTracking: true
      }
    ]);
```

In the orchestrator, I can now use the provider definition to ingest an instance of FullStoryTelemetryProvider by asking the injector for that token, then use it in the same way. There's no benefit to a string lookup here - it's just an extra level of indirection on either side. The caller now needs to provide both the specific instance config, and define the mapping of string to provider; the orchestrator needs to get that specific instance config and go in reverse, finding that mapping to get the provider.

Reply:
Makes sense. Can you help me understand how would you associate `FullStoryTelemetryProvider` token with its corresponding class impl `FullStoryTelemetry`?

Reply:
I think just confusion in terminology here - in angular a provider is two things: a token to get an instance (as in, what a dependent give the injector when requesting injection), and a factory to build one (what the injector needs to actually build an instance for that token). Most of the time, we use a class constructor as both rolled into one, and that's the context I'm using the term here (I was calling the class `FullStoryTelemetryProvider`, you were calling it `FullStoryTelemetry`).

In other words, if my class was named `FullStoryTelemetry`, I'd pass in `telemetryProvider: FullStoryTelemetry` as part of my provider definition and the orchestrator would call the injector with `injector.get(providerDef.telemetryProvider)` to get an instance of `FullStoryTelemetry`.

Reply:
Got it. But here is a circular dep. As per the current design, the provider class needs to take a config in its constructor. In the above interface, the provider class is part of the **config**. They need to be separated now. Also, I need to check how we can pass a constructor argument in the injector.get method.

Reply:
okay. saw your other comment. Let me spend some more time over it.